docs/docsite/rst/dev_guide/testing_units_modules.rst

```diff
@@ -0,0 +1,479 @@
+may extend beyond testing only minimal units and in some cases will include some
+level of functional testing.
+
+
+Why (and why not)?
+------------------
+
+Ansible unit tests have advantages and disadvantages. It is important to understand these:
+
+* Most unit tests are much faster than most Ansible integration tests.  
+* Unit tests can be run by developers who don't have access to system which the module
+  (module dependencies aren't required with unit tests)
+
+There are also some potential disadvantages of unit tests. Unit tests don't normally
+test actual useful valuable features of software, instead just internal implementation
+
+* Unit tests that test the internal, non-visible features of software may make
+  refactoring difficult if those internal features have to change (see also naming in How
+  below)
+* Even if the internal feature is working correctly it is possible that there will be a
+  problem between the internal code and the actual result delivered to the user
+
+Normally the Ansible integration tests (which are written in Ansible YAML) are a better
+option for most module tests.  If those tests already test a feature and run acceptably
+quickly then there may be little point in providing a unit test as well.
+
+When (and when not)?
+--------------------
+
+There are a number of situations where unit tests win out over integration tests. For
+example, testing things which are impossible, slow or very difficult to test with
+integration tests, such as:
+    
+* Forcing rare / strange / random situations that can't be forced, such as specific network
+  failures and exceptions
+* Extensive testing of slow configuration APIs 
+
+Example:
+  A single step of the rds_instance test cases can take up to 20
+  minutes (the time to create an RDS instance in Amazon).  The entire
+  test run can last for well over an hour.  All 16 of the unit tests
+  complete execution in less than 2 seconds.
+
+The time a developer can save by being able to run the code in a unit test makes it worth
+creating a unit test to help resolve many of the problems identified in the module even if
+those tests do not often identify problems later. Every module should have at
+least one unit test which will give quick feedback in easy cases without having to wait
+for the integration tests to complete.
+    
+Providing simple external tests which check the module runs the calls that it makes to
+external services in the right way.
+
+Example:
+  Package managers are often far more efficient when installing multiple packages at once
+  rather than each package separately. The final result is the
+  same: the packages are all installed, so the efficiency is difficult to verify through
+  integration tests. By providing a mock package manager and verifying that it is called
+  once, we can build a valuable test for module efficiency.
+
+Providing specific design tests; By building a requirement for a particular part of the
```

Comment:
Did you mean the first part of this paragraph to be a heading? Please clarify.

Reply:
Wasn't meant to be but it's a good idea;  I have now created headers for several sections.